### PR TITLE
user changed

### PR DESCRIPTION
**Pull request by berkyaglioglu**

### CODE REVIEW

File: users/forms.py [hunk 1]
```diff
@@ -1,16 +1,18 @@
...
 
     class Meta:
         model = User
-        fields=['first_name', 'last_name', 'username', 'password']
+        fields = ['first_name', 'last_name', 'gender', 'email', 'password']
+
 
 class LoginForm(forms.Form):
-    username = forms.CharField(max_length=50,required=True, widget=forms.TextInput())
+    email = forms.CharField(max_length=50, required=True, widget=forms.TextInput())
```
Comment (gokselcoban) [inline users/forms.py]: Length of an email can be more than 50. Also, forms  should have email field.

File: users/migrations/0008_auto_20190725_1243.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+# Generated by Django 2.0.6 on 2019-07-25 09:43
+
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('users', '0007_auto_20190724_1816'),
+    ]
```
Comment (gokselcoban) [inline users/migrations/0008_auto_20190725_1243.py:10]: Can you please give meaningful names to new migrations ?

**Review by gokselcoban — CHANGES_REQUESTED** [review]

File: posts/views.py [hunk 3]
```diff
@@ -25,3 +27,24 @@ def post_list(request):
...
+
+
+def handle_like(request, pk):
+    post = Post.objects.get(pk=pk)
+
+    if request.method == 'POST':
+        if request.user in post.likes.all():
+            post.likes.remove(request.user)
+        else:
+            post.likes.add(request.user)
```
Comment (gokselcoban) [inline posts/views.py]: `request.user` can be `None`

File: templates/posts/detail_user.html [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+<html lang="en">
+<head>
+    <link rel="stylesheet" href="{% static 'css/global.css'%}" type="text/css">
+    <meta charset="UTF-8">
+    <title>Recipes</title>
+</head>
+<body>
+<div style="margin: 0 auto;width: 500px">
+    {% if request.user.is_authenticated %}
```
Comment (gokselcoban) [inline templates/posts/detail_user.html]: Why did you use two different `html` file ? If you are sure that the user is authenticated, why did you check `request.user.is_authenticated` ?

File: users/forms.py [hunk 5]
```diff
@@ -1,16 +1,18 @@
...
 
     class Meta:
         model = User
-        fields=['first_name', 'last_name', 'username', 'password']
+        fields = ['first_name', 'last_name', 'gender', 'email', 'password']
+
 
 class LoginForm(forms.Form):
-    username = forms.CharField(max_length=50,required=True, widget=forms.TextInput())
+    email = forms.CharField(max_length=100, required=True, widget=forms.TextInput())
```
Comment (gokselcoban) [inline users/forms.py]: Let's use `forms.EmailField` .

File: users/forms.py [hunk 6]
```diff
@@ -1,16 +1,18 @@
 from django import forms
-from django.contrib.auth.models import User
+from .models import User
+
 
 class Register(forms.ModelForm):
```
Comment (gokselcoban) [inline users/forms.py]: `Register` is not a clear name. Please use `RegistrationForm` or any other clear name.

File: users/forms.py [hunk 7]
```diff
@@ -1,16 +1,18 @@
...
     class Meta:
         model = User
-        fields=['first_name', 'last_name', 'username', 'password']
+        fields = ['first_name', 'last_name', 'gender', 'email', 'password']
+
 
 class LoginForm(forms.Form):
-    username = forms.CharField(max_length=50,required=True, widget=forms.TextInput())
+    email = forms.CharField(max_length=100, required=True, widget=forms.TextInput())
     password = forms.CharField(max_length=50, widget=forms.PasswordInput())
```
Comment (gokselcoban) [inline users/forms.py]: What is the max_length of password on database ?

File: templates/posts/detail_anonymous.html [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+<html lang="en">
+<head>
+    <link rel="stylesheet" href="{% static 'css/global.css'%}" type="text/css">
+    <meta charset="UTF-8">
+    <title>Recipes</title>
+</head>
+<body>
+<div style="margin: 0 auto;width: 500px">
+    {% if request.user.is_authenticated %}
```
Comment (gokselcoban) [inline templates/posts/detail_anonymous.html]: As far as I can see, there is no `request` in `context`. Is it added automatically ?

`return render(request, 'posts/detail_anonymous.html', context={'post': post})`
https://github.com/berkyaglioglu/Post-Sharing-Application/pull/7/files#diff-7f5d848739da3904921a61c635a3254bR36

File: posts/views.py (hunk 3 above)
Comment (gokselcoban) [reply posts/views.py]: I can open the same recipe on two tabs at the same time. Then, I can click to like button on two different tabs. What is the expected result ?

File: posts/urls.py [hunk 9]
```diff
@@ -6,4 +6,6 @@
 urlpatterns=[
     path('', view=views.post_list, name='post_list'),
     path('create/', view=views.post_create, name='post_create'),
+    path('detail/<int:pk>/', view=views.post_detail, name='detail'),
```
Comment (gokselcoban) [inline posts/urls.py]: `posts/detail/<id>/` should be `posts/1/`

File: posts/models.py [hunk 10]
```diff
@@ -5,7 +5,9 @@
 
 class Post(models.Model):
 
-    # user = models.ForeignKey(User, on_delete=models.CASCADE)
+    user = models.ForeignKey(User, on_delete=models.CASCADE, related_name='user')
+    likes = models.ManyToManyField(User, related_name='likes')
```
Comment (gokselcoban) [inline posts/models.py]: Can you get when a like is created ?

Let's add separate model `Like` and store date created.

File: templates/posts/detail_anonymous.html [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    {% else %}
+    <h2><a href="{% url 'posts:post_list' %}"> RECIPES </a></h2>
+    <h2><a href="{% url 'users:register' %}"> REGISTER </a></h2>
+    <h2><a href="{% url 'users:login' %}"> LOGIN </a></h2>
+    {% endif %}
+</div>
+<hr>
+<h2>{{ post.title }}</h2>
+<p> by {{ post.user.first_name }} {{ post.user.last_name }} - {{ post.created_date }} </p>
+<img src="{{ post.img.url }}">
```
Comment (gokselcoban) [inline templates/posts/detail_anonymous.html]: Please do not use abbreviation for field names and rename `img` to `image`.
Also rename `created_date` to `date_created` and `updated_date` to `date_updated`

**Review by gokselcoban — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 12]
```diff
@@ -5,5 +5,6 @@ users/__pycache__/*
 .pyc
 .idea/workspace.xml
 *.pyc
```
Comment (gokselcoban) [inline .gitignore:7]: Let's add media folder to gitignore.

File: posts/migrations/0003_auto_20190729_1545.py [hunk 13]
```diff
@@ -0,0 +1,22 @@
+# Generated by Django 2.0.6 on 2019-07-29 12:45
+
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('posts', '0002_auto_20190729_1442'),
```
Comment (gokselcoban) [inline posts/migrations/0003_auto_20190729_1545.py]: You should give meaningful names to your migration files.

File: posts/models.py [hunk 14]
```diff
@@ -1,17 +1,36 @@
...
+        ('celery', 'Celery'),
+        ('egg', 'Egg'),
+        ('milk', 'Milk'),
+        ('fish', 'Fish'),
+        ('chicken', 'Chicken'),
+        ('oil', 'Oil')
+    )
+    # picked_ingredients
+    ingredients = models.ManyToManyField(Ingredients)
+    choices = MultiSelectField(choices=INGREDIENT_CHOICES, max_length=1000, blank=True)
```
Comment (gokselcoban) [inline posts/models.py]: You should not add `choices` field. You should list existing `Ingredients` and user select these.
If you list this static list to users, when a new ingredient object is added you should update your code/list.

File: posts/models.py [hunk 15]
```diff
@@ -20,3 +39,21 @@ class Meta:
...
+    post = models.ForeignKey(Post, on_delete=models.CASCADE)
+
+    date_created = models.DateTimeField(auto_now_add=True)
+
+
+class Rate(models.Model):
+    user = models.ForeignKey(User, on_delete=models.CASCADE)
+    post = models.ForeignKey(Post, on_delete=models.CASCADE)
+
+    RATINGS = ((1, '1'), (2, '2'), (3, '3'), (4, '4'), (5, '5'))
```
Comment (gokselcoban) [inline posts/models.py]: Let's rename to `RATINGS_CHOICES`.

File: posts/models.py [hunk 16]
```diff
@@ -20,3 +39,21 @@ class Meta:
...
+    rating = models.IntegerField(max_length=1, choices=RATINGS, default=5)
```
Comment (gokselcoban) [inline posts/models.py]: We can use `score` instead of `rating`. What do you think ? `rating` in `Rate` model is not good.

Comment (gokselcoban) [inline posts/models.py]: I think you should not give a default value for rating. It is error prone. For any reason you can forgot to pass actual rating value.

File: posts/views.py [hunk 17]
```diff
@@ -16,12 +17,150 @@ def post_create(request):
     if request.method == 'POST':
         post_form = PostForm(request.POST, request.FILES)
         if post_form.is_valid():
-            post_form.save(commit=True)
+            post = post_form.save(commit=False)
+            post.owner = request.user
+            post.save()
+            for choice in post.choices:
```
Comment (gokselcoban) [inline posts/views.py]: You can get ingredients or ingredient ids from the user. You should update this part after removing `choices`.

File: posts/views.py [hunk 18]
```diff
@@ -16,12 +17,150 @@ def post_create(request):
...
 
-def post_list(request):
-    post_list = Post.objects.all()
-    return render(request, 'posts/post_list.html', context={'post_list': post_list})
+def post_list(request, pk=None):
+    if pk is None:
+        post_list = Post.objects.all()
+    else:
+        ingredient = Ingredients.objects.get(pk=pk)
+        post_list = ingredient.post_set.all()
```
Comment (gokselcoban) [inline posts/views.py]: Post list view should not get `pk`. Your post detail page should get `pk`.

File: posts/models.py [hunk 19]
```diff
@@ -1,17 +1,36 @@
...
+        ('eggplant', 'Eggplant'),
+        ('celery', 'Celery'),
+        ('egg', 'Egg'),
+        ('milk', 'Milk'),
+        ('fish', 'Fish'),
+        ('chicken', 'Chicken'),
+        ('oil', 'Oil')
+    )
+    # picked_ingredients
+    ingredients = models.ManyToManyField(Ingredients)
```
Comment (gokselcoban) [inline posts/models.py]: You can add `related_name`. `related_name="posts"` and use `posts` to access post of ingredients -> `ingredient.posts` `ingredients.posts.all()`
https://docs.djangoproject.com/en/2.2/ref/models/fields/#django.db.models.ManyToManyField.related_name

File: posts/views.py [hunk 20]
```diff
@@ -16,12 +17,150 @@ def post_create(request):
...
+
+    ingredient_counts = {}
+    for ingredient in Ingredients.objects.all():
+        ingredient_counts[ingredient.name] = ingredient.post_set.count()
+    sorted_ingredient_counts = sorted(ingredient_counts.items(), key=operator.itemgetter(1))
+    sorted_ingredient_counts.reverse()
+    ingredients_list = []
+    for ingredient in sorted_ingredient_counts:
+        ingredients_list.append(Ingredients.objects.get(name=ingredient[0]))
```
Comment (gokselcoban) [inline posts/views.py]: Sorry, I cannot understand this part. From line 39 to 46.

File: posts/views.py [hunk 21]
```diff
@@ -16,12 +17,150 @@ def post_create(request):
...
+
+    for post in post_list:
+        num = 0
+        for rate in post.rate_set.all():
+            num += rate.rating
+        if post.rate_set.all():
```
Comment (gokselcoban) [inline posts/views.py]: You should use `exists()` for this kind of controls.   `post.rate_set.all().exists()`

File: posts/views.py (hunk 20 above)
Comment (berkyaglioglu) [reply posts/views.py]: It creates a dictionary where the keys are ingredient names, and the values are the numbers indicating how many posts that the specified ingredient is in. Then, it sorts the dictionary according to values, and reverse the sorted list to get the list as from the most used one to the least used one.  Then, it creates a list just out of ingredient names in descending order of their usages.

File: posts/models.py [hunk 22]
```diff
@@ -1,17 +1,36 @@
...
+    choices = MultiSelectField(choices=INGREDIENT_CHOICES, max_length=1000, blank=True)
 
-    # user = models.ForeignKey(User, on_delete=models.CASCADE)
+    rating = models.FloatField(max_length=10, default=0)
```
Comment (gokselcoban) [inline posts/models.py]: You should not store average rating in `Post` model. Let's remove `rating` field. It is similar to `like count`, we talked about it with @yigitguler.

File: posts/views.py [hunk 23]
```diff
@@ -16,12 +17,150 @@ def post_create(request):
...
+            average = num / post.rate_set.all().count()
+            post.rating = average
+        else:
+            post.rating = 0
+        post.save()
```
Comment (gokselcoban) [inline posts/views.py]: We can calculate average rating count on database level with using `annotate`.
https://docs.djangoproject.com/en/2.2/topics/db/aggregation/

Current implementation sends to many requests to database.

File: posts/views.py (hunk 20 above)
Comment (gokselcoban) [reply posts/views.py]: Thank you. Let's do this calculation on database.
1 - annotate the `post_count`
2 - order `Ingredients` by `post_count`

File: posts/views.py [hunk 24]
```diff
@@ -16,12 +17,150 @@ def post_create(request):
...
+        post = Post.objects.get(pk=pk)
+
+        if post.owner == request.user:
+            post.delete()
+
+    return HttpResponseRedirect(reverse('posts:post_list'))
+
+
+def handle_like(request, pk):
+    if request.user.is_authenticated:
```
Comment (gokselcoban) [inline posts/views.py]: Can we use `login-required` decorator in this view and in other required views ?
https://docs.djangoproject.com/en/2.2/topics/auth/default/#the-login-required-decorator

File: posts/views.py [hunk 25]
```diff
@@ -16,12 +17,150 @@ def post_create(request):
...
+            return render(request, 'posts/post_edit.html', context={'form': form, 'pk': pk})
+
+    return HttpResponseRedirect(reverse('posts:detail', kwargs={'pk': pk}))
+
+
+def delete(request, pk):
+    if request.user.is_authenticated and request.method == 'POST':
+        post = Post.objects.get(pk=pk)
+
+        if post.owner == request.user:
```
Comment (cobang) [inline posts/views.py]: We can raise `PermissionDenied` error if the user is not owner.

File: users/forms.py (hunk 7 above)
Comment (gokselcoban) [reply users/forms.py]: @berkyaglioglu Did you check ?

File: posts/forms.py [hunk 26]
```diff
@@ -1,13 +1,23 @@
...
 
     class Meta:
         model = Post
-        fields = ['title', 'ingredients', 'description', 'img']
+        fields = ['title', 'description', 'image', 'ingredients']
+
 
+class RateForm(forms.ModelForm):
+    def __init__(self, *args, **kwargs):
+        super(RateForm, self).__init__(*args, **kwargs)
```
Comment (gokselcoban) [inline posts/forms.py]: Looks like you do not need to override `__init__` method.

File: posts/urls.py [hunk 27]
```diff
@@ -6,4 +6,11 @@
 urlpatterns=[
     path('', view=views.post_list, name='post_list'),
     path('create/', view=views.post_create, name='post_create'),
-]
\ No newline at end of file
+    path('<int:pk>/', view=views.post_detail, name='detail'),
+    path('like/<int:pk>/', view=views.handle_like, name='like'),
```
Comment (gokselcoban) [inline posts/urls.py]: You should update the URLs. For example, `posts/like/<post_id>/` should be `posts/<post_id>/like/`

File: posts/views.py [hunk 28]
```diff
@@ -1,27 +1,145 @@
...
+            post.ingredients.set(post_form.cleaned_data['ingredients'])
             return HttpResponseRedirect(reverse('posts:post_list'))
     return render(request, 'posts/post_create.html', context={'form': post_form})
 
 
 def post_list(request):
-    post_list = Post.objects.all()
-    return render(request, 'posts/post_list.html', context={'post_list': post_list})
+    if request.GET.get('ingredient', None) is None:
+        post_list = Post.objects.annotate(avg_rating=Avg('rate__score'))
```
Comment (gokselcoban) [inline posts/views.py]: post_list is not `list` as a data type. It is a bit confusing. Let's rename to `posts` or `post_queryset`

File: posts/views.py [hunk 29]
```diff
@@ -1,27 +1,145 @@
...
+    else:
+        ingredient = Ingredients.objects.get(pk=request.GET.get('ingredient'))
```
Comment (gokselcoban) [inline posts/views.py]: User can enter `?ingredient=123asd123` . Let's handle this issue.

File: posts/views.py [hunk 30]
```diff
@@ -1,27 +1,145 @@
...
+    return render(request, 'posts/post_list.html', context={'post_list': post_list, 'ingredients': ingredients})
+
+
+def post_detail(request, pk):
+    post = Post.objects.get(pk=pk)
+    rate_form = RateForm()
+
+    if request.user.is_authenticated:
+        like = Like.objects.filter(post=post, user=request.user)
+        if Rate.objects.filter(post=post, user=request.user):
```
Comment (gokselcoban) [inline posts/views.py]: Let's use `exists`

File: posts/views.py [hunk 31]
```diff
@@ -1,27 +1,145 @@
...
+
+    return HttpResponseRedirect(reverse('posts:post_list'))
+
+
+def handle_like(request, pk):
+    if request.user.is_authenticated:
+        post = Post.objects.get(pk=pk)
+
+        if request.method == 'POST':
+            if not Like.objects.filter(post=post, user=request.user):
```
Comment (gokselcoban) [inline posts/views.py]: `exists`

File: posts/views.py [hunk 32]
```diff
@@ -1,27 +1,145 @@
...
+
+    return HttpResponseRedirect(reverse('posts:detail', kwargs={'pk': pk}))
+
+
+def handle_unlike(request, pk):
+    if request.user.is_authenticated:
+        post = Post.objects.get(pk=pk)
+
+        if request.method == 'POST':
+            if Like.objects.filter(post=post, user=request.user):
```
Comment (gokselcoban) [inline posts/views.py]: `exists`

File: posts/urls.py [hunk 33]
```diff
@@ -7,10 +7,10 @@
     path('', view=views.post_list, name='post_list'),
     path('create/', view=views.post_create, name='post_create'),
     path('<int:pk>/', view=views.post_detail, name='detail'),
+    path('<int:pk>/edit/', view=views.post_edit, name='edit'),
+    path('<int:pk>/delete/', view=views.post_delete, name='delete'),
+    path('<int:pk>/update/', view=views.handle_update, name='update'),
     path('<int:pk>/like/', view=views.handle_like, name='like'),
```
Comment (gokselcoban) [inline posts/urls.py]: `handle_like` -> `post_like`

File: requirements.txt [hunk 34]
```diff
@@ -0,0 +1 @@
+Django 2.2.1
```
Comment (gokselcoban) [inline requirements.txt:1]: Let's check the format of `requirements.txt`

**Pull request merged**